passes/opt/share.cc

```diff
@@ -1299,45 +1346,57 @@ struct ShareWorker
 					other_cell_active.push_back(qcsat.ez->vec_eq(qcsat.importSig(p.first), qcsat.importSig(p.second)));
 					all_ctrl_signals.append(p.first);
 				}
+				int sub1 = qcsat.ez->expression(qcsat.ez->OpOr, cell_active);
+				int sub2 = qcsat.ez->expression(qcsat.ez->OpOr, other_cell_active);
 
+				bool pattern_only_solve = qcsat.ez->solve(qcsat.ez->AND(sub1, sub2));
 				qcsat.prepare();
 
-				int sub1 = qcsat.ez->expression(qcsat.ez->OpOr, cell_active);
 				if (!qcsat.ez->solve(sub1)) {
 					log("      According to the SAT solver the cell %s is never active. Sharing is pointless, we simply remove it.\n", log_id(cell));
 					cells_to_remove.insert(cell);
 					break;
 				}
 
-				int sub2 = qcsat.ez->expression(qcsat.ez->OpOr, other_cell_active);
 				if (!qcsat.ez->solve(sub2)) {
 					log("      According to the SAT solver the cell %s is never active. Sharing is pointless, we simply remove it.\n", log_id(other_cell));
 					cells_to_remove.insert(other_cell);
 					shareable_cells.erase(other_cell);
 					continue;
 				}
 
-				qcsat.ez->non_incremental();
+				if (pattern_only_solve) {
```

Comment:
Isn't this condition inverted? The "then" branch is pre-existing code, now used for when there is no solution found before importing the design. When I invert this, a simple example I cooked up starts working. It's not enough to pass tests/sat but it seems necessary

Reply:
Oh it actually is sufficient to pass the tests, I just got pranked by the `Expected error pattern 'Signal difference' found !!!` line again. It's so mean

Reply:
No, the condition is correct because it's true if not mutually exclusive, so there has to be a different reason

Reply:
If `pattern_only_solve` is true, that means there was a counterexample to mutual exclusion when just considering the patterns. That's the case where we do want to fall back to the previous behavior, as they might still be mutually exclusive when we re-solve with the input cones imported, but in that case we can't optimize the activation patterns.